### PR TITLE
[Flax T5] Speed up t5 training

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

This PR makes sure that no jax functionality is used in the preprocessing to make sure the TPU is not unnecessarly blocked. This small change leads to a **5x** factor speed-up in training T5.

🚨🚨 **Note**: It is extremely important to verify that no DeviceArrays are created during the data preprocessing to make sure that the training step can run asynchronously on TPU while the preprocessing runs on CPU. A good rule is to make sure that in the training step look, only the function `p_train_step` uses JAX/Flax code and all other functions run on CPU. Other relevant links: https://jax.readthedocs.io/en/latest/async_dispatch.html#async-dispatch 🚨🚨

### CODE REVIEW

File: examples/flax/language-modeling/run_t5_mlm_flax.py [hunk 1]
```diff
@@ -353,7 +353,8 @@ def _random_segmentation(num_items, num_segments):
             np.random.shuffle(mask_indices)
             first_in_segment = np.pad(mask_indices, [[1, 0]])
             segment_id = np.cumsum(first_in_segment)
-            segment_length = np.asarray(jax.ops.segment_sum(np.ones_like(segment_id), segment_id))
```
Comment (patrickvonplaten) [inline examples/flax/language-modeling/run_t5_mlm_flax.py:356]: Removing jax from the data collator leads to a huge speed-up here (from 2s/iteration to 1.5 iterations / s)

**Comment by patrickvonplaten** [comment]
The PR is tested here: https://huggingface.co/patrickvonplaten/t5-base-norwegian/tensorboard (check train loss graph which shows that time is reduced to < 5h now)

File: examples/flax/language-modeling/README.md [hunk 2]
```diff
@@ -215,6 +215,42 @@ config = GPT2Config.from_pretrained("gpt2", resid_pdrop=0.0, embd_pdrop=0.0, att
 config.save_pretrained(model_dir)
 ```
 
+### Train tokenizer
```
Comment (patrickvonplaten) [inline examples/flax/language-modeling/README.md:218]: This wasn't added in a previous PR - correcting it here.

File: examples/flax/language-modeling/README.md [hunk 3]
```diff
@@ -373,15 +409,15 @@ Next we can run the example script to pretrain the model:
...
+	--save_steps="10000" \
+	--eval_steps="2500" \
 	--push_to_hub
 ```
 
 Training should converge at a loss and accuracy 
-of 2.2 and 58.0 respectively after 2 epochs on a single TPUv3-8.
-This should take around 24 hours.
+of 2.36 and 57.0 respectively after 3 epochs on a single TPUv3-8.
+This should take around 4.5 hours.
```
Comment (patrickvonplaten) [inline examples/flax/language-modeling/README.md:420]: yaaay

**Review by sgugger — APPROVED** [review]
Nice!
